File: bu_utils/populate_db.js
import county_codes from './assets/county_codes.json';
import axios from 'axios';
import fs from 'fs';
import path from 'path';

const BU_FILES_PATH = './results';
const BU_TREE_PREFIX = 'bu_tree_';
const BACKEND_URL = 'http://localhost:8080';


const send_bu_to_backend = async (tree, bu) => {
    axios.post(BACKEND_URL + '/bu/create', {
        "tree_name": tree,
        "data": bu
    })
}

const create_tree = async (tree_name, commitment_size) => {
    axios.post(BACKEND_URL + '/tree/create_tree', {
        "tree_name": tree_name,
        "commitment_size": commitment_size
    })
}

const commit_tree = async (tree_name) => {
    axios.post(BACKEND_URL + '/tree/commit', {
        "tree_name": tree_name
    })
}

const bu_files_path = () => {
    try {
        const files = fs.readdirSync(bu_files_path);
        return files.map(file => path.join(bu_files_path, file));
    } catch (err) {
        console.info(`No files found in ${bu_files_path}, reading example files`); 
        return ['./assets/example_bus_consolidated.json'];
    }
}

const send_bus_to_backend = async () => {
    const files = bu_files_path();
    for (const file of files) {
        const bu_file = fs.readFileSync(file);
        const bu = JSON.parse(bu_file);
        const tree_name = BU_TREE_PREFIX + path.basename(file, '.json');
        await create_tree(tree_name, 1);
        await send_bu_to_backend(tree_name, bu);
        await commit_tree(tree_name);
    }

}
    

const parse_bu_for_election = (bu) => {
    /* Common values for each election */
    const zona = bu['identificacaoSecao']['municipioZona']['zona']
    const secao = bu['identificacaoSecao']['secao']
    const cod_municipio = bu['identificacaoSecao']['municipioZona']['municipio']
    const municipio = get_municipio_from_code(cod_municipio)
    const uf = municipio['uf']
    const nome_municipio = municipio['nome_municipio']

    /* Object for each election */
    const elections = []
    for (let election in bu['resultadosVotacaoPorEleicao']) {
        const id_eleicao = election['idEleicao']

        elections.push({
            "id_eleicao": id_eleicao,
            "UF": uf,
            "zona": zona,
            "secao": secao,
            "municipio": nome_municipio,
            "bu_inteiro": bu,
        })
    }
    return elections
}

const get_municipio_from_code = (cod_municipio) => {
    for (const county_code in county_codes) 
        if (county_code === cod_municipio) 
            return county_codes[county_code]

    return {
        "codigo_tse": cod_municipio,
        "uf":"ZZ",
        "nome_municipio":"Exterior",
        "capital": null,
        "codigo_ibge": null
    }
}

File: bu_utils/populate_db.py
import os
import requests

from constants import BACKEND_URL


def commit_all_trees():
    response = requests.post(f'{BACKEND_URL}/tree/commit-all-trees')
    print(response.json())
    return response


def send_file_to_backend(file_path: str):
    print(f"Sending file {file_path} to backend")
    with open(file_path, "rb") as file:
        response = requests.post(f'{BACKEND_URL}/bu/create', files={"file": (os.path.basename(file_path), file)})
        print(response.json())
        return response


def read_bu_or_busa_files():
    files = [os.path.join(root, name)
                for root, dirs, files in os.walk("./assets/bus")
                for name in files if name.endswith((".bu", ".busa"))]
    
    if not files:
        files = os.listdir("./assets/mocked_bus")
        files = [f"./assets/mocked_bus/{file}" for file in files]
        
    return files


def insert_list_bus_to_db():
    for file in read_bu_or_busa_files():
        send_file_to_backend(file)
    print("---- Finished")


if __name__ == '__main__':
    insert_list_bus_to_db()
    commit_all_trees()
